common/raw_hashtable_metadata_group_benchmark.cpp

```diff
@@ -0,0 +1,233 @@
+
+      // Replicate the match byte in each of the other matching indices.
+      for (ssize_t m_index : llvm::ArrayRef(group_indices)
+                                 .drop_front()
+                                 .take_front(match_count - 1)) {
+        group_bytes[m_index] = match_b;
+      }
+    }
+
+    // Now that the storage is set up, record these in our struct.
+    bm.metadata = metadata_storage[bm_index];
+    if constexpr (Kind == BenchKind::Random) {
+      bm.bytes = bytes_storage[bm_index];
+    }
+  }
+  return bm_storage;
+}
+
+template <BenchKind Kind>
+// NOLINTNEXTLINE(google-readability-casting): False positive clang-tidy bug.
+const auto bench_metadata = BuildBenchMetadata<Kind>();
+
+template <BenchKind Kind, typename GroupT = MetadataGroup>
+static void BM_LoadMatch(benchmark::State& s) {
+  BenchMetadata bm = bench_metadata<Kind>[0];
+  ssize_t nonce_data[GroupSize];
+  for (ssize_t& n : nonce_data) {
+    n = 1;
+  }
+  // Now hide the contents of `nonce_data`.
+  benchmark::ClobberMemory();
+  ssize_t i = 0;
+  benchmark::DoNotOptimize(i);
+#pragma clang loop unroll(disable)
+  for (auto _ : s) {
+    auto g = GroupT::Load(bm.metadata.data(), i * GroupSize);
+    typename GroupT::MatchIndex matches;
+    if constexpr (Kind == BenchKind::Empty) {
+      matches = g.MatchEmpty();
+    } else if constexpr (Kind == BenchKind::Deleted) {
+      matches = g.MatchDeleted();
+    } else {
+      static_assert(Kind == BenchKind::Random);
+      matches = static_cast<MetadataGroup::MatchIndex>(g.Match(bm.bytes[i]));
+    }
+    CARBON_CHECK(matches);
+    i = (i + nonce_data[matches.index()]) & (BenchSize - 1);
+  }
+}
+BENCHMARK(BM_LoadMatch<BenchKind::Random>);
+BENCHMARK(BM_LoadMatch<BenchKind::Empty>);
+BENCHMARK(BM_LoadMatch<BenchKind::Deleted>);
+#if CARBON_NEON_SIMD_SUPPORT || CARBON_X86_SIMD_SUPPORT
+BENCHMARK(BM_LoadMatch<BenchKind::Random, PortableGroup>);
+BENCHMARK(BM_LoadMatch<BenchKind::Empty, PortableGroup>);
+BENCHMARK(BM_LoadMatch<BenchKind::Deleted, PortableGroup>);
+BENCHMARK(BM_LoadMatch<BenchKind::Random, SIMDGroup>);
+BENCHMARK(BM_LoadMatch<BenchKind::Empty, SIMDGroup>);
+BENCHMARK(BM_LoadMatch<BenchKind::Deleted, SIMDGroup>);
+#endif
+
+template <BenchKind Kind, ssize_t Steps>
+static void BM_LoadMatchMissSteps(benchmark::State& s) {
+  static_assert(Steps > 0);
+  static_assert(Steps <= GroupSize);
+  BenchMetadata bm = bench_metadata<Kind>[Steps - 1];
+  ssize_t nonce_data[GroupSize];
+  for (ssize_t& n : nonce_data) {
+    n = 1;
+  }
+  // Now hide the contents of `nonce_data`.
+  benchmark::ClobberMemory();
+  ssize_t i = 0;
+  benchmark::DoNotOptimize(i);
+#pragma clang loop unroll(disable)
+  for (auto _ : s) {
+    auto g = MetadataGroup::Load(bm.metadata.data(), i * GroupSize);
+    auto matched_range = g.Match(bm.bytes[i]);
+    ssize_t indices = 0;
+    for (ssize_t index : matched_range) {
+      indices += nonce_data[index];
+    }
+    i = (i + (indices - Steps + 1)) & (BenchSize - 1);
```

Comment:
This is particularly mysterious

Reply:
Yeah, many comments added. Hopefully helps.